README.md
# MRS simulation

| Build status | [![Build Status](https://github.com/ctu-mrs/mrs_simulation/workflows/Melodic/badge.svg)](https://github.com/ctu-mrs/mrs_simulation/actions) | [![Build Status](https://github.com/ctu-mrs/mrs_simulation/workflows/Noetic/badge.svg)](https://github.com/ctu-mrs/mrs_simulation/actions) |
|--------------|---------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------|

## Overview
Support for spawning vehicles into Gazebo simulation, where user can select from multiple UAV platforms.
This platforms can be additionaly equipped with several sensors (rangefinders, 2d lidars, stereo cameras etc.).

## Usage

> :warning: **New drone spawning mechanism**:
>
> * From now on, the drone spawning mechanism will be controlled by a ROS node
> * Spawning new vehicles is now done by calling ROS services
> * The spawner node is launched automatically with the simulation
> * The arguments for vehicle configuration remain mostly unchanged, just use 'rosservice call /mrs_drone_spawner/spawn "args"' instead of 'spawn_uav args'
> * Distributed simulation (running gazebo on one computer and mavros/PX4 on another is currently not supported


### Start of the Gazebo simulator

To start the prepared example of Gazebo world call:

```bash
roslaunch mrs_simulation simulation.launch world_file:='$(find mrs_gazebo_common)/worlds/grass_plane.world' gui:=true
```

At this point the Gazebo world will only contain the environment with grass plane but with no vehicles yet.

### Spawning of UAVs ([NEW](https://ctu-mrs.github.io/docs/simulation/drone_spawner.html))
The `simulation.launch` will automatically start the `mrs_drone_spawner` python node. If you use a custom launch file to start the simulation, you can start it separately:

```bash
roslaunch mrs_simulation mrs_drone_spawner.launch
```

The `mrs_drone_spawner` will perform the following tasks:

* Spawn vehicle models in the Gazebo simulation (ids from 0 to 250). This is done internally by calling the command `rosrun gazebo_ros spawn_model`.

* For each vehicle, PX4 firmware and mavros is started at specific port numbers depending on the vehicle ID.

Vehicles are added to the simulation by calling the `spawn` service of the `mrs_drone_spawner`. The service takes one string argument, which specifies the vehicle ID, type and sensor configuration. Example: spawn a single vehicle with a down-facing laser rangefinder:

```bash
rosservice call /mrs_drone_spawner/spawn "1 --enable-rangefinder"
```

To display the manual containing a list of all available arguments, perform a dry-run of the script:

```bash
rosrun mrs_simulation mrs_drone_spawner
```

The arguments are also listed in the `mrs_simulation/config/spawner_params.yaml` file. Note that not all sensors are available for all the vehicle types. The config file stores the available configurations in the following format: `parameter: [default_value, help_description, [compatible_vehicles]]`

#### New features
Multiple vehicles may be spawned with one service call:

```bash
rosservice call /mrs_drone_spawner/spawn "1 2 3 4 5 --t650 --enable-bluefox-camera --enable-rangefinder"
```

Spawn position may be specified by a command line argument `--pos x y z heading`: [m, m, m, rad]
```bash
rosservice call /mrs_drone_spawner/spawn "1 --f550 --enable-rangefinder --pos 10 -15 0.3 0.7"
```
For multiple vehicles, `--pos` defines the spawn point of the first vehicle. Following vehicles will be spawned in a line, with an x-offset of 2 meters from the previous vehicle.

Spawn position may be specified by a `.csv` or a `.yaml` file using `--file absolute_path_to_file`:
```bash
rosservice call /mrs_drone_spawner/spawn "1 --f450 --enable-rangefinder --enable-ouster --use-gpu-ray --ouster-model OS1-64 --file `pwd`/spawn_poses.yaml"
```

Use a whitespace instead of an ID to get an available ID automatically assigned by the spawner.
```bash
rosservice call /mrs_drone_spawner/spawn " --f450 --enable-rangefinder"
```

<!-- THIS SECTION IS OUTDATED
Not all sensors have to be available for selected type of uav (DJI f450, DJI f550, Tarot t650 and Eagle.one mk2). Please check possible settings for
the selected type of UAV by calling command:

```bash
spawn_uav --$UAV_TYPE --available-sensors
```

#### Spawning on single PC

An example for spawning the DJI F550 vehicle with id 1 and additional sensors, running generated launch file directly, and deleting models from gazebo simulation after closing the script.

```bash
spawn_uav 1 --f550 --enable-bluefox-camera --enable-rangefinder --run --delete
```

To spawn multiple vehicles use additional terminal windows/panels or chain the spawn commands by running the previous commands in background.

```bash
spawn_uav 1 --f550 --run --delete --enable-rangefinder & spawn_uav 2 --f550 --run --delete --enable-rangefinder && fg
```

Note that this approach will hide the output of all but one spawn script. A recommended practise to using multiple terminal panels.

#### Distributed simulation
In order to run a simulation distributed to multiple machines, these prerequisities have to be met:
* A central machine with IP address `IP_ADDR_MASTER` must be running ROS master, i.e. perform `roslaunch mrs_simulation simulation.launch` to start Gazebo.
* Following variables have to be set in the environment on all used machines. On a machine with IP address `IP_ADDR`, add following lines to the `~/.bashrc` script (do not forget to source the script after).

```bash
    export ROS_IP=IP_ADDR
    export ROS_MASTER_URI=http://IP_ADDR_MASTER:11311
    export GAZEBO_IP=IP_ADDR
    export GAZEBO_MASTER_URI=http://IP_ADDR_MASTER:12345
```

* To spawn a vehicle from non-central machine with IP address `IP_ADDR`, extend the spawning example with parameter `--mavlink-address`, i.e.

```bash
    spawn_uav 1 --enable-bluefox-camera --enable-rangefinder --run --delete --mavlink-address IP_ADDR
```
-->

### Running the MRS control pipeline
Check out the wiki on [how to run control core](https://ctu-mrs.github.io/docs/simulation/howto.html#3-run-the-control-core).


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package mrs_simulation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.3 (2022-05-09)
------------------
* add front-facing realsense to x500
* updated mbzirc sensor package
* refactored mavros launchfiles
* + install in cmakelists
* partial merge of marine branch
* Fix roll limits
* Update servo camera xacros
* update naki servo camera daes and their origins
* naki: split gimbal from frame to enable movable visual
* Update servo camera macro
  Add NAKI II drone to simulation
* ouster diag topic defaults now to os_cloud_nodelet/is_alive to match hw driver
* fix naki propellers rotation
* add safety led macro
* add macro for ultrasonic sensors; add them to NAKI drone
* add models/visuals/sensor setups for NAKI drone
* fixed whycon box for f450
* update motor constants
  random
* initial commit with naki files: copy of f450, modeless, octo_coax
* new realsense loading, fixed multiple instances
* Revert "updating realsense naming"
  This reverts commit 96156a3f6dbff2c2a64157790c8bb8c64ec1204c.
* updating realsense naming
* random
* rotated vio IMU to correspond with the real drone
* add timepix3 sensor to x500 and t650 (`#18 <https://github.com/ctu-mrs/mrs_simulation/issues/18>`_)
* added realsense front to f330
* update brus spawner options
* added missing tag to t650
* added enable-light option to x500
* added x500 ouster support
* add brus to spawner (`#17 <https://github.com/ctu-mrs/mrs_simulation/issues/17>`_)
* added ground truth plugin to all uavs, added custom mag plugin
* Contributors: Dan Hert, DanHert, Matej Petrlik, Pavel Petracek, Pavel Petráček, Tomas Baca, Vaclav Pritzl, Vit Kratky, stibipet

1.0.1 (2021-05-16)
------------------
* increased timeouts in drone spawner to 60s
* add rad120 to component_snippets.xacro
* Adding the third/back camera to simulation
* enabled ouster for eaglemk2
* Merge branch 'uvdar_better_signalling'
* added realsense_up_down parameter for 2 realsense cameras pointing up and down
* Fixing the camera frametate at 60Hz
* Adding signal_id parameter to component snippets
* Changing the frequency parameter to signal id parameter
* Contributors: Matouš Vrba, Pavel Petracek, Tomas Baca, Vaclav Pritzl, Viktor Walter, ViktorWalter

1.0.0 (2021-03-18)
------------------
* Major release


scripts/simulation/spawn.py
import argparse
import subprocess
import sys

from simulation import get_launch_snippet
from simulation import get_vehicle_pose
from simulation import get_vehicle_pose_from_file
from simulation import delete_model
from simulation import spawn_model
from simulation import get_all_vehicle_ports
from simulation import write_launch_file
from simulation import print_ok
from simulation import detect_available_arguments_in_xacro 
from simulation import get_model_xacro_file 


def vehicle_id_type(value):
    """Validate the vehicle_id_type from string to int."""
    value = int(value)
    if value < 0 or value > 250:
        raise argparse.ArgumentTypeError('Vehicle id must be in [1, 250]')
    return value

def spawn():
    parser = argparse.ArgumentParser(description = 'Spawn vehicles into gazebo simulation.', 
                                     epilog = """Not all sensors have to be available for selected type of uav.
                                     Please check possible settings for the selected type of UAV by calling command: spawn --$UAV_TYPE --available_sensors""",
                                     formatter_class=lambda prog: argparse.HelpFormatter(prog,max_help_position=27, width= 100))
    parser.add_argument(
        'vehicle_id', nargs='*', metavar='VEHICLE_ID', type=vehicle_id_type,
        default=(1,),
        help='The id of a vehicle to spawn, when the load file is not specified (default: 1)')
    type_group = parser.add_mutually_exclusive_group()
    type_group.add_argument(
        '--f450', action = 'store_true',
        help='Change the spawned vehicle to DJI F450 (default is DJI F550).'
    )
    type_group.add_argument(
        '--f550',action = 'store_true',
        help='Force the spawned vehicle to DJI F550.'
    )
    type_group.add_argument(
        '--t650', action = 'store_true',
        help='Change the spawned vehicle to Tarot 650 (default is DJI F550)'
    )
    type_group.add_argument(
        '--eaglemk2', action = 'store_true',
        help='Change the spawned vehicle to Eagle.One Mk2 (default is DJI F550).'
    )
    parser.add_argument(
        '--available-sensors', action = 'store_true',
        help='Display all available sensor equipment that is possible to use for selected type of uav.'
    )
    parser.add_argument(
        '--enable-ground-truth', action = 'store_true',
        help='Enable topic with ground true odometry (default: false)')
    parser.add_argument(
        '--run', action='store_true',
        help='run generated launch file (default: false)')
    parser.add_argument(
        '--delete', action='store_true',
        help='Despawn when killed (default: false)')
    parser.add_argument(
        '--file',
        help='Load positions and ids from \'*.csv\' file with format: [id, x, y, z, heading] or \'*.yaml\' file with format: [uav_name: \\newline id: (int) \\newline x: (float) \\newline y: (float) \\newline z: (float) \\newline heading: (float)]')
    bluefox_down_group = parser.add_mutually_exclusive_group()
    bluefox_down_group.add_argument(
        '--enable-bluefox-camera', action ='store_true',
        help='Add bluefox camera to the vehicle [752x480 50hz] (default: false)')
    bluefox_down_group.add_argument(
        '--enable-bluefox-camera-reverse', action ='store_true',
        help='Add bluefox camera to the vehicle [752x480 50hz], rotated by 180 deg (default: false)')
    parser.add_argument(
        '--enable-fisheye-camera', action ='store_true',
        help='Add fisheye camera to the vehicle [752x480 180deg 45hz] (default: false)')
    parser.add_argument(
        '--enable-whycon-box', action ='store_true',
        help='Add whycon box for relative localization (default: false)')
    parser.add_argument(
        '--enable-mobius-camera-down', action = 'store_true',
        help='Add mobius camera to the vehicle [1280x720 30hz], pointed to ground (default: false)')
    parser.add_argument(
        '--enable-mobius-camera-front', action = 'store_true',
        help='Add mobius camera to the vehicle [1280x720 30hz], pointed to the front (default: false)')
    parser.add_argument(
        '--enable-mobius-camera-back-left', action = 'store_true',
        help='Add mobius camera to the vehicle [1280x720 30hz], pointed to the back left (default: false)')
    parser.add_argument(
        '--enable-mobius-camera-back-right', action = 'store_true',
        help='Add mobius camera to the vehicle [1280x720 30hz], pointed to the back right (default: false)')
    realsense_group = parser.add_mutually_exclusive_group()
    realsense_group.add_argument(
        '--enable-realsense-front-pitched', action = 'store_true',
        help='Add Intel Realsense D435 depth camera to the vehicle [1280x720 30hz], pointed to the front pitched down by 10 up to 45 deg. Depends on the type of vehicle (default: false)')
    realsense_group.add_argument(
        '--enable-realsense-down', action = 'store_true',
        help='Add Intel Realsense D435 depth camera to the vehicle [1280x720 30hz], pointed down (default: false)')
    realsense_group.add_argument(
        '--enable-realsense-top', action = 'store_true',
        help='Add Intel Realsense D435 depth camera to the vehicle [1280x720 30hz], pointed forward placed on the aluminum frame (default: false)')
    realsense_group.add_argument(
        '--enable-realsense-front', action = 'store_true',
        help='Add Intel Realsense D435 depth camera to the vehicle [1280x720 30hz], pointed forward placed on the front holder between the legs (default: false)')
    realsense_group.add_argument(
        '--enable-realsense-up', action = 'store_true',
        help='Add Intel Realsense D435 depth camera to the vehicle [1280x720 30hz], pointed upwards (default: false)')
    parser.add_argument(
        '--use-realistic-realsense', action = 'store_true',
        help='Enable a more realistic simulation of the Realsense D435 depth image (default: false, only takes effect if some Realsense is enabled)')
    parser.add_argument(
        '--enable-rangefinder', action = 'store_true',
        help='Add rangefinder to the vehicle (default: false)')
    parser.add_argument(
        '--enable-teraranger', action = 'store_true',
        help='Add teraranger rangefinder to the vehicle (default: false)')
    parser.add_argument(
        '--enable-rangefinder-up', action = 'store_true',
        help='Add rangefinder measuring upwards to the vehicle (default: false)')
    parser.add_argument(
        '--enable-gripper', action = 'store_true',
        help='Add magnetic gripper (default: false)')
    lidar_group = parser.add_mutually_exclusive_group()
    lidar_group.add_argument(
        '--enable-scanse', action = 'store_true',
        help='Add the Scanse Sweep laser scanner to the vehicle (default: false)')
    lidar_group.add_argument(
        '--enable-rplidar', action = 'store_true',
        help='Add the RPlidar A3 laser scanner to the vehicle (default: false)')
    lidar_group.add_argument(
        '--enable-teraranger-tower-evo', action = 'store_true',
        help='Add the Teraranger Tower Evo laser scanner to the vehicle (default: false)')
    lidar_group.add_argument(
        '--enable-velodyne', action = 'store_true',
        help='Add the Velodyne PUCK Lite laser scanner to the vehicle (default: false)')
    lidar_group.add_argument(
        '--enable-ouster', action = 'store_true',
        help='Add the Ouster laser scanner to the vehicle (default: false)')
    parser.add_argument(
        '--ouster-model', nargs=1, type=str, 
        default=('OS1-16'), 
        choices =['OS0-32', 'OS0-64', 'OS0-128', 'OS1-16', 'OS1-32G1', 'OS1-32', 'OS1-64', 'OS1-128', 'OS2-32', 'OS2-64', 'OS2-128'], 
        help='Choose the Ouster model (default: %(default)s)')
    parser.add_argument(
        '--use-gpu-ray', action = 'store_true',
        help='Laser ray casting will be handled by GPU shaders (default: false)')
    parser.add_argument(
        '--enable-timepix', action = 'store_true',
        help='Add Timepix detector to the drone (default: false)')
    parser.add_argument(
        '--enable-thermal-camera', action = 'store_true',
        help='Add thermal camera to the drone (default: false)')
    parser.add_argument(
        '--enable-ball-holder', action = 'store_true',
        help='Add the ball holder to the vehicle (default: false)')
    parser.add_argument(
        '--enable-light', action = 'store_true',
        help='Add the light to the vehicle (default: false)')
    parser.add_argument(
        '--enable-servo-camera', action = 'store_true',
        help='Add the camera on virtual servo to the vehicle (default: false)')
    parser.add_argument(
        '--enable-water-gun', action = 'store_true',
        help='Add water gun for fire fighting (default: false)')
    parser.add_argument(
        '--enable-parachute', action = 'store_true',
        help='Add a parachute for emergency flight termination (default: false)')
    parser.add_argument(
        '--gps-indoor-jamming', action = 'store_true',
        help='Enable gps jamming when drone is indoors (default:false)')
    parser.add_argument(
        '--enable-pendulum', action = 'store_true',
        help='Add pendulum to the drone (default: false)')
    parser.add_argument(
        '--enable-uv-leds', action = 'store_true',
        help='Add UV LEDs on the vehicle (default: false)')
    parser.add_argument(
        '--led-frequencies', nargs=2, type=int, default=(6,15),
        help='Specify UV LEDs frequencies (default: (6,15))')
    parser.add_argument(
        '--enable-uv-leds-beacon', action = 'store_true',
        help='Add UV LED beacon on the top of the vehicle (default: false)')
    parser.add_argument(
        '--beacon-frequency', nargs=1, type=int, default=[30],
        help='The frequency of blinking of the UV Beacon')
    parser.add_argument(
        '--enable-uv-camera', action = 'store_true',
        help='Add UV camera on the vehicle (default: false)')
    parser.add_argument(
        '--uv-camera-calibration-file', nargs=1, type=str, default=("~/.ros/calib_results.txt"),
        help='Specify UV camera calibration different than default one')
    parser.add_argument(
        '--use-uv-occlusions', action = 'store_true',
        help='Enable occlusions for UVDAR simulation - heavy on computation (default: false)')
    mbzirc_group = parser.add_mutually_exclusive_group()
    mbzirc_group.add_argument(
        '--wall-challenge', action = 'store_true',
        help='Configuration for the MBZIRC wall challenge (T650 with 2x bluefox down, realsense down pitched, rangefinder down) (default: false)')
    mbzirc_group.add_argument(
        '--fire-challenge', action = 'store_true',
        help='Configuration for the MBZIRC fire challenge (T650 with bluefox down, realsense forward, rp lidar, 3x thermal camera, water gun) (default: false)')
    mbzirc_group.add_argument(
        '--fire-challenge-blanket', action = 'store_true',
        help='Configuration for the MBZIRC fire challenge (blanket dropping) (T650 with bluefox down, realsense forward, rp lidar, 2x thermal camera, maybe blanket dropper) (default: false)')
    parser.add_argument(
        '--gazebo-ros-master-uri',
        help='The uri used to spawn the models')
    parser.add_argument(
        '--mavlink-address',
        help='The IP address for mavlink (default: INADDR_ANY)')
    parser.add_argument(
        '--generate-launch-file', action='store_false',
        help='Generate launch file (default: true)')
    parser.add_argument('--debug', action='store_true',
        help='Print generated xml model file (default: false)')

    args = parser.parse_args()

    vehicle_id = args.vehicle_id

    n_vehicles = len(vehicle_id)
    # do not remove functionality (it may get fixed someday), just clip the nubmer of vehicles to 1
    if (n_vehicles > 1):
        print('\033[31;1m' + '[WARN]: Spawning multiple vehicles in one command is currently not supported.\n[WARN]: Only the first vehicle will spawn.\n' + '\033[32;1m' + '[HINT]: Create a separate tmux pane (ctrl+s) and call the script again to spawn more vehicles' + '\x1b[0m')
        n_vehicles = 1
    
    if args.f450:
        vehicle_type='f450'
    elif args.f550:
        vehicle_type='f550'
    elif args.t650:
        vehicle_type='t650'
    elif args.eaglemk2:
        vehicle_type='eaglemk2'
    else:
        vehicle_type='f550'

    if args.wall_challenge or args.fire_challenge or args.fire_challenge_blanket:
        vehicle_type='t650'

    if args.fire_challenge:
        args.enable_rangefinder = True
        args.enable_rangefinder_up = True
        args.enable_rplidar = True
        args.enable_realsense_front = True
        args.enable_thermal_camera = True
        args.enable_water_gun = True

    if args.available_sensors:
        _, model_xml = get_model_xacro_file(vehicle_type)
        for setting in detect_available_arguments_in_xacro(model_xml):
            print('--{}'.format(setting.replace('_','-')))
        return

    if args.file:
        vehicle_id, pose = get_vehicle_pose_from_file(args.file, vehicle_id[0])
        n_vehicles = len(vehicle_id)

    ouster_model = args.ouster_model[0].upper()

    uvled_fr = args.led_frequencies
    uvled_beacon_fr = args.beacon_frequency
    uvcam_calib = args.uv_camera_calibration_file

    if args.enable_uv_camera:
        print("UV camera calibration file is set to %s" %args.uv_camera_calibration_file)
    
    launch_snippet = ''

    # spawn vehicles
    if n_vehicles == 1:
        print("[MrsSimulation]: Spawning vehicle!")
    else:
        print("[MrsSimulation]: Spawning %i vehicles!" %n_vehicles)

    for i in range(0,n_vehicles):

        mav_sys_id = vehicle_id[i]

        vehicle_ports = get_all_vehicle_ports(mav_sys_id)

        if args.file:
            vehicle_pose = poses[i]
        else:
            vehicle_pose = get_vehicle_pose(mav_sys_id)
        
        # spawn the vehicle model in gazebo
        spawn_model(
            mav_sys_id = mav_sys_id,
            vehicle_type = vehicle_type, 
            mavlink_udp_port = vehicle_ports['mavlink_udp_port'], 
            mavlink_tcp_port = vehicle_ports['mavlink_tcp_port'],
            pose = vehicle_pose,
            ros_master_uri = args.gazebo_ros_master_uri,
            mavlink_address = args.mavlink_address,
            enable_rangefinder = args.enable_rangefinder,
            enable_teraranger = args.enable_teraranger,
            enable_rangefinder_up = args.enable_rangefinder_up,
            enable_ground_truth = args.enable_ground_truth,
            enable_mobius_camera_front = args.enable_mobius_camera_front,
            enable_mobius_camera_down = args.enable_mobius_camera_down,
            enable_mobius_camera_back_left = args.enable_mobius_camera_back_left,
            enable_mobius_camera_back_right = args.enable_mobius_camera_back_right,
            enable_realsense_front_pitched = args.enable_realsense_front_pitched,
            enable_realsense_down = args.enable_realsense_down,
            enable_realsense_top = args.enable_realsense_top,
            enable_realsense_front = args.enable_realsense_front,
            enable_realsense_up = args.enable_realsense_up,
            use_realistic_realsense = args.use_realistic_realsense,
            enable_whycon_box = args.enable_whycon_box,
            enable_bluefox_camera = args.enable_bluefox_camera,
            enable_bluefox_camera_reverse = args.enable_bluefox_camera_reverse,
            enable_fisheye_camera = args.enable_fisheye_camera,
            enable_magnetic_gripper = args.enable_gripper,
            enable_scanse_sweep = args.enable_scanse,
            enable_pendulum = args.enable_pendulum,
            enable_rplidar = args.enable_rplidar,
            enable_teraranger_tower_evo = args.enable_teraranger_tower_evo,
            enable_timepix = args.enable_timepix,
            enable_thermal_camera = args.enable_thermal_camera,
            enable_velodyne = args.enable_velodyne,
            enable_ouster = args.enable_ouster,
            ouster_model = ouster_model,
            use_gpu_ray = args.use_gpu_ray,
            enable_light = args.enable_light,
            enable_servo_camera = args.enable_servo_camera,
            enable_ball_holder = args.enable_ball_holder,
            enable_water_gun = args.enable_water_gun,
            enable_parachute = args.enable_parachute,
            wall_challenge = args.wall_challenge,
            fire_challenge_blanket = args.fire_challenge_blanket,
            gps_indoor_jamming = args.gps_indoor_jamming,
            enable_uv_leds = args.enable_uv_leds,
            uvled_fr_l = "{}".format(uvled_fr[0]),
            uvled_fr_r = "{}".format(uvled_fr[1]),
            enable_uv_leds_beacon = args.enable_uv_leds_beacon,
            uvled_beacon_f = "{}".format(uvled_beacon_fr[0]),
            enable_uv_camera = args.enable_uv_camera,
            uvcam_calib_file = uvcam_calib[0].strip(),
            uvcam_occlusions = args.use_uv_occlusions,
            debug=args.debug)

        if args.generate_launch_file or args.run:
            launch_snippet += get_launch_snippet(mav_sys_id, vehicle_type)
    print("----------------------------------------------")

    if args.generate_launch_file or args.run:
        launch_path = write_launch_file(launch_snippet)
        cmd = ['roslaunch', launch_path]
        print_ok(' '.join(cmd))

    retcode = 0
    if args.run:
        try:
            retcode = subprocess.call(cmd)
        except KeyboardInterrupt:
            pass
    if args.delete:
        for i in range(0, n_vehicles):
            mav_sys_id = vehicle_id[i]
            delete_model(
                mav_sys_id,
                vehicle_type,
                ros_master_uri=args.gazebo_ros_master_uri)
    return retcode
